File: src/main/java/komu/blunt/types/checker/TypeChecker.java
package komu.blunt.types.checker;

import komu.blunt.ast.AST;
import komu.blunt.ast.ASTExpression;
import komu.blunt.ast.ASTValueDefinition;
import komu.blunt.ast.BindGroup;
import komu.blunt.eval.TypeCheckException;
import komu.blunt.types.*;
import komu.blunt.types.patterns.Pattern;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;
import static komu.blunt.types.Qualified.quantifyAll;
import static komu.blunt.types.Type.typeVariable;
import static komu.blunt.types.checker.Unifier.mgu;

public final class TypeChecker {

    public final ClassEnv classEnv;
    private final DataTypeDefinitions dataTypes;
    private int typeSequence = 0;
    private final ExpressionTypeCheckVisitor expressionVisitor = new ExpressionTypeCheckVisitor(this);
    private final BindingTypeChecker bindingTypeChecker = new BindingTypeChecker(this);
    private final PatternTypeChecker patternTypeChecker = new PatternTypeChecker(this);
    private Substitution substitution = Substitution.empty();

    private TypeChecker(ClassEnv classEnv, DataTypeDefinitions dataTypes) {
        this.classEnv = checkNotNull(classEnv);
        this.dataTypes = checkNotNull(dataTypes);
    }

    public static Qualified<Type> typeCheck(ASTExpression exp, ClassEnv classEnv, DataTypeDefinitions dataTypes, Assumptions as) {
        TypeChecker checker = new TypeChecker(classEnv, dataTypes);

        return checker.normalize(checker.typeCheck(exp, as));
    }

    public static Scheme typeCheck(ASTValueDefinition exp, ClassEnv classEnv, DataTypeDefinitions dataTypes, Assumptions as) {
        TypeChecker checker = new TypeChecker(classEnv, dataTypes);

        return quantifyAll(checker.normalize(checker.typeCheck(exp, as)));
    }

    private Qualified<Type> normalize(TypeCheckResult<Type> result) {
        List<Predicate> ps = classEnv.reduce(applySubstitution(result.predicates));
        return applySubstitution(new Qualified<>(ps, result.value));
    }

    TypeCheckResult<Type> typeCheck(ASTExpression exp, Assumptions as) {
        return exp.accept(expressionVisitor, as);
    }

    TypeCheckResult<Assumptions> typeCheckBindGroup(BindGroup bindGroup, Assumptions as) {
        return bindingTypeChecker.typeCheckBindGroup(bindGroup, as);
    }

    TypeCheckResult<Type> typeCheck(ASTValueDefinition define, Assumptions as) {
        ASTExpression let = AST.letRec(define.name, define.value, AST.variable(define.name));
        return typeCheck(let, as);
    }

    PatternTypeCheckResult<Type> typeCheck(Pattern pattern) {
        return pattern.accept(patternTypeChecker, null);
    }

    Qualified<Type> freshInstance(Scheme scheme) {
        List<TypeVariable> ts = new ArrayList<>(scheme.kinds.size());
        for (Kind kind : scheme.kinds)
            ts.add(newTVar(kind));

        return Qualified.instantiate(ts, scheme.type);
    }

    TypeVariable newTVar() {
        return newTVar(Kind.STAR);
    }

    TypeVariable newTVar(Kind kind) {
        return typeVariable(typeName(typeSequence++), kind);
    }

    List<Type> newTVars(int size) {
        List<Type> types = new ArrayList<>(size);
        for (int i = 0; i < size; i++)
            types.add(newTVar());
        return types;
    }

    private static String typeName(int index) {
        if (index < 5) {
            return String.valueOf((char) ('a' + index));
        } else {
            return "t" + (index-5);
        }
    }

    void unify(Type t1, Type t2) {
        try {
            Substitution u = mgu(t1.apply(substitution), t2.apply(substitution));
            substitution = u.compose(substitution);
        } catch (UnificationException e) {
            throw new TypeCheckException(e);
        }
    }

    <T extends Types<T>> T applySubstitution(T t) {
        return t.apply(substitution);
    }

    <T extends Types<T>> List<T> applySubstitution(Collection<T> ts) {
        return TypeUtils.applySubstitution(substitution, ts);
    }

    public ConstructorDefinition findConstructor(String name) {
        return dataTypes.findConstructor(name);
    }
}

File: src/main/kotlin/komu/blunt/types/checker/TypeChecker.kt
package komu.blunt.types.checker

import komu.blunt.ast.ASTDefinition
import komu.blunt.ast.ASTExpression
import komu.blunt.ast.BindGroup
import komu.blunt.eval.TypeCheckException
import komu.blunt.types.*
import komu.blunt.types.patterns.Pattern

class TypeChecker(val classEnv: ClassEnv, private val dataTypes: DataTypeDefinitions) {

    private var typeSequence = 0
    private val expressionVisitor = ExpressionTypeCheckVisitor(this)
    private val bindingTypeChecker = BindingTypeChecker(this)
    private val patternTypeChecker = PatternTypeChecker(this)
    private var substitution = Substitution.empty

    companion object {

        fun typeCheck(exp: ASTExpression, classEnv: ClassEnv, dataTypes: DataTypeDefinitions, ass: Assumptions): Qualified<Type> {
            val checker = TypeChecker(classEnv, dataTypes)
            return checker.normalize(checker.typeCheck(exp, ass))
        }

        fun typeCheck(exp: ASTDefinition.Value, classEnv: ClassEnv, dataTypes: DataTypeDefinitions, ass: Assumptions): Scheme {
            val checker = TypeChecker(classEnv, dataTypes)
            return checker.normalize(checker.typeCheck(exp, ass)).quantifyAll()
        }
    }

    private fun normalize(result: TypeCheckResult<Type>): Qualified<Type> {
        val ps = classEnv.reduce(applySubstitution(result.predicates))
        return Qualified(ps, result.value).apply(substitution)
    }

    fun typeCheck(exp: ASTExpression, ass: Assumptions): TypeCheckResult<Type> =
        expressionVisitor.typeCheck(exp, ass)

    fun typeCheckBindGroup(bindGroup: BindGroup, ass: Assumptions): TypeCheckResult<Assumptions> =
        bindingTypeChecker.typeCheckBindGroup(bindGroup, ass)

    fun typeCheck(define: ASTDefinition.Value, ass: Assumptions): TypeCheckResult<Type> =
        typeCheck(ASTExpression.LetRec(define.name, define.value, ASTExpression.Variable(define.name)), ass)

    fun typeCheck(pattern: Pattern): PatternTypeCheckResult<Type> =
        patternTypeChecker.typeCheck(pattern)

    fun freshInstance(scheme: Scheme): Qualified<Type> =
        scheme.type.instantiate(newTVars(scheme.kinds))

    fun newTVar(): Type.Var =
        newTVar(Kind.Star)

    fun newTVar(kind: Kind): Type.Var =
        Type.Var(typeName(typeSequence++), kind)

    fun newTVars(size: Int): List<Type.Var> =
        (1..size).map { newTVar() }

    fun newTVars(kinds: List<Kind>): List<Type.Var> =
        kinds.map { newTVar(it) }

    private fun typeName(index: Int): String =
        if (index < 5)
            ('a' + index).toChar().toString()
        else
            "t" + (index-5)

    fun unify(t1: Type, t2: Type) {
        val tt1 = t1.apply(substitution)
        val tt2 = t2.apply(substitution)
        try {
            val u = Unifier.mgu(tt1, tt2)
            substitution = u.compose(substitution)
        } catch (e: UnificationException) {
            throw TypeCheckException("$tt1 -- $tt2 -- $e")
        }
    }

    // TODO
    fun applySubstitution(t: Assumptions): Assumptions =
      t.apply(substitution)

    fun <T : Types<T>> applySubstitution(ts: Collection<T>): List<T> =
        ts.map { it.apply(substitution) }

    fun findConstructor(name: String): ConstructorDefinition =
        dataTypes.findConstructor(name)
}
